python/addons/preprocessors.py

```diff
@@ -0,0 +1,175 @@
+            if os.path.exists(feature_vocab_file):
+                indices[feature], vocabs[feature] = self._read_vocab(feature_vocab_file, feature)
+        return indices, vocabs
+
+    @staticmethod
+    def create_vectorizers(model_base_dir, pid):
+        """
+        :model_file the path-like object to the model and model name.
+        :vocab_suffixes the list of vocab types. e.g. 'word', 'char', 'ner'.
+        """
+        import pickle
+        return pickle.load(open(os.path.join(model_base_dir, "vectorizers-{}.pkl".format(pid)), "rb"))
+
+    def _assign_char_lookup(self):
+        upchars = tf.constant([chr(i) for i in range(65, 91)])
+        self.lchars = tf.constant([chr(i) for i in range(97, 123)])
+        self.upchars_lut = tf.contrib.lookup.index_table_from_tensor(mapping=upchars, num_oov_buckets=1, default_value=-1)
+
+    def _create_example(self):
+        serialized_tf_example = tf.placeholder(tf.string, name='tf_example')
+        feature_configs = {
+            self.FIELD_NAME: tf.FixedLenFeature(shape=[], dtype=tf.string),
+        }
+        tf_example = tf.parse_example(serialized_tf_example, feature_configs)
+        return tf_example
+
+    def run(self):
+        self._assign_char_lookup()
+        tf_example = self._create_example()
+        preprocessed = self.create_preprocessed_input(tf_example)
+        return tf_example, preprocessed
+
+    def create_preprocessed_input(self, tf_example):
+        """
+        Create a preprocessor chain inside of the tensorflow graph.
+        """
+        pass
+
+    @staticmethod
+    def reshape_indices(indices, shape):
+        reshaped = tf.sparse_reset_shape(indices, new_shape=shape)
+        # Now convert to a dense representation
+        x = tf.sparse_tensor_to_dense(reshaped)
+        x = tf.contrib.framework.with_shape(shape, x)
+        return x
+
+
+class Token1DPreprocessorCreator(PreprocessorCreator):
+
+    def __init__(self, model_base_dir, pid, features, **kwargs):
+        super(Token1DPreprocessorCreator, self).__init__(model_base_dir, pid, features, **kwargs)
+        self.mxlen = self.vectorizers['word'].mxlen
+        self.FIELD_NAME = 'tokens'
+
+    def reform_raw(self, raw):
+        """
+        Splits and rejoins a string to ensure that tokens meet
+        the required max len.
+        """
+        raw_tokens = tf.string_split(tf.reshape(raw, [-1])).values
+        # sentence length <= mxlen
+        raw_post = tf.reduce_join(raw_tokens[:self.mxlen], separator=" ")
+        return raw_post
+
+    def create_word_vectors_from_post(self, raw_post):
+        # vocab has only lowercase words
+        word2index = self.indices['word']
+        split_chars = tf.string_split(tf.reshape(raw_post, [-1]), delimiter="").values
+        upchar_inds = self.upchars_lut.lookup(split_chars)
+        lc_raw_post = tf.reduce_join(tf.map_fn(lambda x: tf.cond(x[0] > 25,
```

Comment:
Shouldn't this be conditional? there are times when people don't want to lowercase their input